provider/agent.go

```diff
@@ -22,10 +25,38 @@ func agentResource() *schema.Resource {
 		SchemaVersion: 1,
 
 		Description: "Use this resource to associate an agent.",
-		CreateContext: func(_ context.Context, resourceData *schema.ResourceData, i interface{}) diag.Diagnostics {
-			// This should be a real authentication token!
-			resourceData.SetId(uuid.NewString())
-			err := resourceData.Set("token", uuid.NewString())
+		CreateContext: func(ctx context.Context, resourceData *schema.ResourceData, i interface{}) diag.Diagnostics {
+			agentID := uuid.NewString()
+			resourceData.SetId(agentID)
+
+			// Most of the time, we will generate a new token for the agent.
+			// In the case of a prebuilt workspace being claimed, we will override with
+			// an existing token provided below.
+			token := uuid.NewString()
+
+			// If isPrebuild is true, then this workspace was built by the prebuilds system.
+			// This does not determine whether the workspace has been claimed by a user.
+			// At this point, it may or may not have been claimed.
+			isPrebuild := helpers.OptionalEnv(IsPrebuildEnvironmentVariable()) == "true"
+			// existingToken should only have been set if isPrebuild is true, because we only
+			// reuse the token when a prebuilt workspace is being claimed.
+			existingToken := helpers.OptionalEnv(RunningAgentTokenEnvironmentVariable(agentID))
+			logFields := map[string]interface{}{
+				"agent_id":       agentID,
+				"is_prebuild":    isPrebuild,
+				"token_provided": existingToken != "",
+			}
+			if isPrebuild && existingToken != "" {
+				// check if a token was already generated for this agent.
+				// If so, this workspace is in the process of being claimed
+				// and we should reuse the token. If not, we use a new token as usual.
+				tflog.Info(ctx, "using provided agent token for prebuild", logFields)
+				token = existingToken
+			} else {
+				tflog.Info(ctx, "using a new agent token", logFields)
+			}
```

Comment:
I don't think it's necessary to log this case.

```suggestion
			}
```